src/persist/src/operators/stream.rs

```diff
@@ -702,19 +709,73 @@ where
                             );
                             let (data, ts, diff) = update;
                             let anti_update = (data, ts, -diff);
-                            write
-                                .write(&[anti_update])
-                                .await
-                                .expect("error persisting retraction");
+                            to_retract.push(anti_update);
                             continue;
                         }
                         session.give(update);
                     }
+
+                    let write_future = write.write(&to_retract);
+                    to_retract.clear();
+
+                    // We are not using the capability for the main output later, but we are
+                    // holding on to it to keep the frontier from advancing because that frontier
+                    // is used downstream to track how far we have persisted. This is used, for
+                    // example, by upsert()/persist()/seal()/conditional_seal() operators
+                    // and await_frontier().
+                    pending_futures.push_back((
+                        cap.delayed(cap.time()),
```

Comment:
Theres two things here that I don't totally get:

1. why are we delaying the cap to its time in line 727? My understanding is that that is a no-op. what am i missing?
2. why do we need to specifically retain the capability for the error output port? (I think there is probably a good reason, I just don't know what it is)

Reply:
And then a 3rd, is this vulnerable to the same kind of issue we found in the conditional_seal operator, where holding onto a cap releases times up to the cap, even though the operation we are waiting on (in this case a write) maybe was not completed yet?

Reply:
>  1. why are we delaying the cap to its time in line 727? My understanding is that that is a no-op. what am i missing?

It's to get around a restriction in the interface, `retain()` consumes `self` while `delayed()` takes a `&self`. But I need to have two capabilities.

>   2. why do we need to specifically retain the capability for the error output port? (I think there is probably a good reason, I just don't know what it is)

That's just how timely works, which I learned it the hard way. :sweat_smile: When you go and try to send something on an output that doesn't match the capability you will get a panic. (c.f. here's where that is checked: https://github.com/timelydataflow/timely-dataflow/blob/b619080c6c3d9ec249b2943d56ff4a2d8ce8e121/timely/src/dataflow/operators/generic/handles.rs#L220)

> And then a 3rd, is this vulnerable to the same kind of issue we found in the conditional_seal operator, where holding onto a cap releases times up to the cap, even though the operation we are waiting on (in this case a write) maybe was not completed yet?

I _think_ it is correct but I have been wrong before. My reasoning is this:

- With `seal`, letting the frontier progress to `t` signals that data with `t'` not beyond `t` can be processed downstream, So letting it advance to `t` before we have successfully written that seal to persistence is wrong.
- With `persist` (and its cousin _filter_and_retract_future_updates_, yes naming is not optimal...) holding the frontier at `t` while we write data at time `t` prevents the downstream frontier from advancing past `t`, which prevents downstream operators from working on the data. For example, `await_frontier()` will only release data with time `t` into the rest of dataflow once `t` is not beyond the frontier anymore.

In other words, `seal` has this "t+1-y" nature while `persist` has a "t-y" nature. Not sure that helps... I'm also partially hoping that you can understand this and convince me back. :sweat_smile:

Reply:
gotcha. For 1. i was expecting to see `cap.retain()` or something like that but it makes sense given the current api why its not possible.

for 3. i think the point of confusion for me is that its not clear to me where the frontiers are and where the data is. I think your example suggests that when we write a batch of records with (data, time, diff), the frontier is held back to a time <= all the times for all the records we are writing, aka the writes are all not beyond the frontier we are stashing. I totally believe thats true, I just don't see where thats enforced. In contrast it seems like the seal frontier is somehow leading indicator -- I need to think about the distinction a bit more

Reply:
For posterity: we discussed this "offline" and came to the conclusion that this should work.

Side information: in our case, it's the `replay()` operator that controls what capabilities are used for sending data and what (differential) timestamp they have.